dispatcher/src/main/java/io/zeebe/dispatcher/Dispatcher.java

```diff
@@ -191,10 +198,12 @@ private long offer(
 
         newPosition = updatePublisherPosition(activePartitionId, newOffset);
 
-        if (publisherPosition.proposeMaxOrdered(newPosition)) {
-          LOG.trace("Updated publisher position to {}", newPosition);
+        // if successful, replace position with simple counter
+        if (newPosition > 0) {
+          newPosition = recordPosition;
+          recordPosition += fragmentCount;
```

Comment:
If we remove the synchronize and make the recordPosition an atomicLong, we have a race condition between claiming the fragments and getting/updating the recordPosition. We didn't have to do this previously because by claiming the fragments we got the position implicitly. Without synchronizing, two concurrent threads can claim two fragments and then the second fragment gets the an earlier position and the first gets a later position.

Reply:
we can use a lock-free approach, which looks like this 

```
do {
currentPosition =   position.get()
nextPosition += fragmentCount;
 } while (!position.cas(currentPosition, nextPosition))
```

Example https://github.com/zeebe-io/zeebe/blob/develop/util/src/main/java/io/zeebe/util/sched/channel/ActorConditions.java#L33

Reply:
Follow up issue to investigate the performance impact: https://github.com/zeebe-io/zeebe/issues/4774